src/gba/io.c

```diff
@@ -343,247 +343,256 @@ void GBAIOInit(struct GBA* gba) {
 void GBAIOWrite(struct GBA* gba, uint32_t address, uint16_t value) {
 	if (address < REG_SOUND1CNT_LO && (address > REG_VCOUNT || address < REG_DISPSTAT)) {
 		value = gba->video.renderer->writeVideoRegister(gba->video.renderer, address, value);
-	} else {
-		switch (address) {
-		// Video
-		case REG_DISPSTAT:
-			value &= 0xFFF8;
-			GBAVideoWriteDISPSTAT(&gba->video, value);
-			return;
+		gba->memory.io[address >> 1] = value;
+		return;
+	}
 
-		case REG_VCOUNT:
-			mLOG(GBA_IO, GAME_ERROR, "Write to read-only I/O register: %03X", address);
+	if (address >= REG_SOUND1CNT_LO && address <= REG_SOUNDCNT_LO) {
+		if (!gba->audio.enable) {
+			// Ignore writes to most audio registers if the hardware is off.
 			return;
+		}
+	}
 
-		// Audio
```

Comment:
I've had to modify the flow of this function to avoid having checks in each case of each sound register. If you check this branch locally with gitk and tick the "ignore whitespace changes" you can see that the actual changes are minimal.